src/ops/custom.rs

```diff
@@ -1,219 +1,186 @@
+    description: String, // cache in advance as description() returns &str
+    args: Vec<TypeArg>,
+    signature: Option<Signature>,
+}
 
-    fn deref(&self) -> &Self::Target {
-        self.custom.as_ref()
-    }
+fn qualify_name(res_id: &ResourceId, op_name: &SmolStr) -> SmolStr {
+    format!("{}/{}", res_id, op_name).into()
 }
 
-impl<T: CustomOp> From<T> for OpaqueOp {
-    fn from(op: T) -> Self {
+impl OpaqueOp {
+    pub fn new(
+        resource: ResourceId,
+        op_name: impl Into<SmolStr>,
+        args: impl Into<Vec<TypeArg>>,
+        signature: Option<Signature>,
+    ) -> Self {
+        let op_name: SmolStr = op_name.into();
+        let description = qualify_name(&resource, &op_name).into();
         Self {
-            id: op.name(),
-            custom: Box::new(op),
+            resource,
+            op_name,
+            description,
+            args: args.into(),
+            signature,
         }
     }
 }
 
-/// Custom definition for an operation.
-///
-/// When implementing this trait, include the `#[typetag::serde]` attribute to
-/// enable serialization.
-#[typetag::serde]
-pub trait CustomOp: Send + Sync + std::fmt::Debug + CustomOpBoxClone + Any + Downcast {
-    /// Try to convert the custom op to a graph definition.
-    ///
-    /// TODO: Create a separate HUGR, or create a children subgraph in the HUGR?
-    fn try_into_hugr(&self, resources: &ResourceSet) -> Option<Hugr> {
-        let _ = resources;
-        None
+impl OpName for ExternalOp {
+    fn name(&self) -> SmolStr {
+        let (res_id, op_name) = match self {
+            Self::Opaque(op) => (&op.resource, &op.op_name),
+            Self::Resource { def, .. } => (&def.resource, &def.name),
+        };
+        qualify_name(res_id, op_name)
     }
+}
 
-    /// List the resources required to execute this operation.
-    fn resources(&self) -> &ResourceSet;
-
-    /// The name of the operation.
-    fn name(&self) -> SmolStr;
-
-    /// Optional description of the operation.
+impl OpTrait for ExternalOp {
     fn description(&self) -> &str {
-        ""
+        match self {
+            Self::Opaque(op) => op.description.as_str(),
+            Self::Resource { def, .. } => def.description.as_str(),
+        }
     }
 
-    /// The signature of the operation.
-    fn signature(&self) -> Signature;
-
-    /// Optional descriptions of the ports in the signature.
     fn signature_desc(&self) -> SignatureDescription {
-        Default::default()
+        match self {
+            Self::Opaque(_) => SignatureDescription::default(),
+            Self::Resource { def, args } => def.signature_desc(args),
+        }
     }
-}
 
-impl_downcast!(CustomOp);
-impl_box_clone!(CustomOp, CustomOpBoxClone);
-
-/// Serializable definition for dynamically loaded operations.
-///
-/// TODO: Define a way to construct new CustomOps from a serialized definition.
-#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
-pub struct OpDef {
```

Comment:
Is OpDef now ExternalOp? This isn't reflected in the spec afaict

Reply:
Op*Def* is now in `resource.rs`. ExternalOp is the instantiation of such a def into a concrete `OpType` for a node.

Reply:
Oops I missed that!